File: skills/hook-creator/scripts/init_hook.py
#!/usr/bin/env python3
"""Add a minimal hook rule and script to an existing plugin directory."""

import argparse
import json
import re
from pathlib import Path

from hook_format import HOOK_EVENTS

NAME_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


def main():
    parser = argparse.ArgumentParser(description="Scaffold a Goose plugin hook")
    parser.add_argument("plugin_dir")
    parser.add_argument("event", choices=sorted(HOOK_EVENTS))
    parser.add_argument("name", help="Kebab-case handler name")
    parser.add_argument("--matcher", default=None)
    parser.add_argument("--timeout", type=int, default=30)
    args = parser.parse_args()

    if not NAME_RE.fullmatch(args.name):
        parser.error("Handler name must be lowercase kebab-case")
    if args.matcher is not None:
        try:
            re.compile(args.matcher)
        except re.error as error:
            parser.error(f"Invalid matcher regex: {error}")
    if args.timeout <= 0:
        parser.error("Timeout must be positive")

    root = Path(args.plugin_dir).expanduser().resolve()
    manifest = root / "plugin.json"
    if not manifest.is_file():
        parser.error(f"Plugin manifest not found: {manifest}")

    hooks_path = root / "hooks" / "hooks.json"
    script_path = root / "scripts" / f"{args.name}.sh"
    if script_path.exists():
        parser.error(f"Refusing to overwrite existing script: {script_path}")

    hooks_path.parent.mkdir(parents=True, exist_ok=True)
    script_path.parent.mkdir(parents=True, exist_ok=True)
    if hooks_path.exists():
        try:
            document = json.loads(hooks_path.read_text(encoding="utf-8"))
        except json.JSONDecodeError as error:
            parser.error(f"Cannot update invalid hooks.json: {error}")
        if not isinstance(document.get("hooks"), dict):
            parser.error("Existing hooks.json must contain a top-level hooks object")
    else:
        document = {"hooks": {}}

    rule = {
        "hooks": [
            {
                "type": "command",
                "command": f"${{PLUGIN_ROOT}}/scripts/{args.name}.sh",
                "timeout": args.timeout,
            }
        ]
    }
    if args.matcher is not None:
        rule["matcher"] = args.matcher
    document["hooks"].setdefault(args.event, []).append(rule)
    hooks_path.write_text(json.dumps(document, indent=2) + "\n", encoding="utf-8")

    script_path.write_text(
        "#!/usr/bin/env sh\n"
        "set -eu\n"
        "payload=$(cat 2>/dev/null || printf '{}')\n"
        "# Implement trusted local automation using the JSON payload.\n"
        "printf '%s' \"$payload\" >/dev/null\n"
        "exit 0\n",
        encoding="utf-8",
    )
    script_path.chmod(script_path.stat().st_mode | 0o111)
    print(hooks_path)
    print(script_path)


if __name__ == "__main__":
    main()

File: skills/hook-creator/scripts/init_hook.ts
#!/usr/bin/env node
// Add a minimal hook rule and script to an existing plugin directory.
import { mkdirSync, existsSync, readFileSync, writeFileSync, chmodSync, statSync } from "node:fs";
import { join, resolve } from "node:path";
import { HOOK_EVENTS } from "./hook_format.js";

const NAME_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

function fail(message: string): never {
  console.error(`error: ${message}`);
  process.exit(2);
  throw new Error("unreachable");
}

function parseArgs(argv: string[]) {
  const positional: string[] = [];
  let matcher: string | null = null;
  let timeout = 30;
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--matcher") {
      matcher = argv[++i] ?? null;
    } else if (arg === "--timeout") {
      timeout = Number(argv[++i]);
    } else if (arg.startsWith("--matcher=")) {
      matcher = arg.slice("--matcher=".length);
    } else if (arg.startsWith("--timeout=")) {
      timeout = Number(arg.slice("--timeout=".length));
    } else {
      positional.push(arg);
    }
  }
  return { positional, matcher, timeout };
}

function main() {
  const { positional, matcher, timeout } = parseArgs(process.argv.slice(2));
  const [pluginDir, event, name] = positional;
  if (!pluginDir || !event || !name) {
    fail("usage: init_hook.js <plugin_dir> <event> <name> [--matcher <regex>] [--timeout <n>]");
  }
  if (!HOOK_EVENTS.has(event)) {
    fail(`invalid event: ${event} (choices: ${Array.from(HOOK_EVENTS).sort().join(", ")})`);
  }
  if (!NAME_RE.test(name)) {
    fail("Handler name must be lowercase kebab-case");
  }
  if (matcher !== null) {
    try {
      new RegExp(matcher);
    } catch (error) {
      fail(`Invalid matcher regex: ${(error as Error).message}`);
    }
  }
  if (!Number.isFinite(timeout) || timeout <= 0) {
    fail("Timeout must be positive");
  }

  const root = resolve(pluginDir);
  const manifest = join(root, "plugin.json");
  if (!existsSync(manifest)) {
    fail(`Plugin manifest not found: ${manifest}`);
  }

  const hooksPath = join(root, "hooks", "hooks.json");
  const scriptPath = join(root, "scripts", `${name}.sh`);
  if (existsSync(scriptPath)) {
    fail(`Refusing to overwrite existing script: ${scriptPath}`);
  }

  mkdirSync(join(root, "hooks"), { recursive: true });
  mkdirSync(join(root, "scripts"), { recursive: true });

  let document: { hooks: Record<string, unknown[]> };
  if (existsSync(hooksPath)) {
    let parsed: unknown;
    try {
      parsed = JSON.parse(readFileSync(hooksPath, "utf-8"));
    } catch (error) {
      fail(`Cannot update invalid hooks.json: ${(error as Error).message}`);
    }
    const candidate = parsed as { hooks?: unknown };
    if (typeof candidate.hooks !== "object" || candidate.hooks === null || Array.isArray(candidate.hooks)) {
      fail("Existing hooks.json must contain a top-level hooks object");
    }
    document = candidate as { hooks: Record<string, unknown[]> };
  } else {
    document = { hooks: {} };
  }

  const rule: Record<string, unknown> = {
    hooks: [
      {
        type: "command",
        command: `\${PLUGIN_ROOT}/scripts/${name}.sh`,
        timeout,
      },
    ],
  };
  if (matcher !== null) {
    rule.matcher = matcher;
  }
  document.hooks[event] = document.hooks[event] ?? [];
  document.hooks[event].push(rule);
  writeFileSync(hooksPath, `${JSON.stringify(document, null, 2)}\n`, "utf-8");

  writeFileSync(
    scriptPath,
    "#!/usr/bin/env sh\n" +
      "set -eu\n" +
      "payload=$(cat 2>/dev/null || printf '{}')\n" +
      "# Implement trusted local automation using the JSON payload.\n" +
      "printf '%s' \"$payload\" >/dev/null\n" +
      "exit 0\n",
    "utf-8"
  );
  const mode = statSync(scriptPath).mode;
  chmodSync(scriptPath, mode | 0o111);

  console.log(hooksPath);
  console.log(scriptPath);
}

main();
